internal/test/e2e/version_compatibility_test.go

```diff
@@ -169,6 +170,102 @@ func TestVersionCompatibility(t *testing.T) {
+	cleanup := run.Koko(t)
+	defer cleanup()
+
+	dpCleanup := run.KongDP(kong.GetKongConfForShared())
+	defer dpCleanup()
+	require.NoError(t, util.WaitForKong(t))
+	require.NoError(t, util.WaitForKongAdminAPI(t))
+
+	kongAdmin, err := kongClient.NewClient(util.BasedKongAdminAPIAddr, nil)
+	require.NoError(t, err, "create go client for kong")
+	ctx := context.Background()
+	info, err := kongAdmin.Root(ctx)
+	require.NoError(t, err, "fetching Kong Gateway info")
+
+	dataPlaneVersion, err := versioning.NewVersion(kongClient.VersionFromInfo(info))
+	require.NoError(t, err)
+
+	admin := httpexpect.New(t, "http://localhost:3000")
+
+	// Remove plugins that may not be expected due to data plane version or aren't subject to compatibility updates
+	expectedPluginsMap := make(map[string]VersionCompatibilityPlugins, 0)
+	for _, pluginTest := range VersionCompatibilityOSSPluginConfigurationTests {
+		addPlugin := true
+		if len(pluginTest.VersionRange) > 0 {
+			version := versioning.MustNewRange(pluginTest.VersionRange)
+			if !version(dataPlaneVersion) {
+				addPlugin = false
+			}
+		}
+		if pluginTest.FieldUpdateChecks == nil {
+			addPlugin = false
+		}
+
+		if addPlugin {
+			expectedPluginsMap[pluginTest.Name] = pluginTest
+		}
+	}
+
+	// create plugin records in the CP + DP
+	for _, pluginTest := range expectedPluginsMap {
+		var config structpb.Struct
+		if len(pluginTest.Config) > 0 {
+			require.Nil(t, json.ProtoJSONUnmarshal([]byte(pluginTest.Config), &config))
+		}
+
+		p := &v1.Plugin{
+			Id:        uuid.NewString(),
+			Name:      pluginTest.Name,
+			Config:    &config,
+			Enabled:   wrapperspb.Bool(true),
+			Protocols: []string{"http", "https"},
+		}
+
+		pluginBytes, err := json.ProtoJSONMarshal(p)
+		require.Nil(t, err)
+		res := admin.POST("/v1/plugins").WithBytes(pluginBytes).Expect()
+		res.Status(http.StatusCreated)
+	}
+
+	// fetch list of created plugin records from the DP
+	dataPlanePlugins := []*kongClient.Plugin{}
+	util.WaitFunc(t, func() error {
+		dataPlanePlugins, err = kongAdmin.Plugins.ListAll(ctx)
+		if err != nil {
+			t.Log("fetch plugin failed", err)
+		}
+		if len(dataPlanePlugins) == 0 {
+			return errors.New("plugins len was 0")
+		}
+		return err
+	})
+
+	dpPluginsMap := make(map[string]*kongClient.Plugin, 0)
+	for _, dpPlugin := range dataPlanePlugins {
+		dpPluginsMap[*dpPlugin.Name] = dpPlugin
+	}
+	require.Equal(t, len(expectedPluginsMap), len(dpPluginsMap), "plugins configured count does not match")
+
+	for name, expectedPlugin := range expectedPluginsMap {
+		t.Run(fmt.Sprintf("%s successfully updates fields", name), func(t *testing.T) {
+			require.Contains(t, dpPluginsMap, name, "failed to discover plugin")
+			for rawVersion, updates := range expectedPlugin.FieldUpdateChecks {
+				parsedVersion := versioning.MustNewRange(rawVersion)
+				if parsedVersion(dataPlaneVersion) {
+					for _, update := range updates {
+						want := fmt.Sprintf("%v", update.Value)
```

Comment:
I think these can simply be (please double-check it!):

```
						want := update.Value
						have := dpPluginsMap[name].Config[update.Field]
```

Reply:
I fumbled around with this for a bit before landing on the presented solution. Here's what happens if I don't stringify and compare:
```
Error:  Not equal: 
            expected: []string([]string{"kong.log.err('Hello Koko!')"})
            actual  : []interface {}([]interface {}{"kong.log.err('Hello Koko!')"})
```

Reply:
I got some feedback from @tjasko as well. Here's the two paths I can take:
```
for _, update := range updates {
	want := update.Value
	have := dpPluginsMap[name].Config[update.Field]
	switch update.Value.(type) {
	case string:
		require.Equal(t, want, have, "failed to validate plugin updates")
	case []string:
		require.ElementsMatch(t, want, have, "failed to validate plugin updates")
	default:
		require.Fail(t, fmt.Sprintf("Unexpected type(s): want: %s, have: %s, ",
		    reflect.TypeOf(want), reflect.TypeOf(have)))
	}
}
```
Or the much less verbose:
```
for _, update := range updates {
	want := fmt.Sprintf("%v", update.Value)
	have := fmt.Sprintf("%v", dpPluginsMap[name].Config[update.Field])
	require.Equal(t, want, have, "failed to validate plugin updates")
}	
```
I'd rather go with the stringified version since these version compatibility handlers have so many conditional statements already. I'm making sure that I'm not introducing something flaky by taking a shortcut.

Reply:
I believe we can happily accept verbosity for correctness. If we are absolutely sure that stringify-all-the-things works well in all cases, including the `[]string` one, then proceed. Otherwise, let's just write some more lines of code. You can always wrap the verification logic in an util function to avoid more nesting:

```
for _, update := range updates {
	want := update.Value
	have := dpPluginsMap[name].Config[update.Field]
	verifyUpdates(want, have) // or something like `require.NoError(t, verifyUpdates(want, have))`
}
```

Reply:
>  can't we simply convert the values (expected and response) to JSON and compare that way?

I considered this, but this is also converting to a string so I would (lazily) just skip the marshaling steps.

I pushed an update that checks the type instead.

Reply:
Your call, but this is going to be a maintenance issue in the future as we will need to keep adding to the case statement for each new type added.

Leaving final decision for @GGabriele.

Reply:
Yeah I agree, I think you can easily re-use part of [the old logic](https://github.com/Kong/koko/pull/368/files#diff-5351ee7e6ddda0e9a38eee6852f369ac1c1fd272bba60c61f6ced419486215c3L456)...something like (definitely needs to be refactored):

```
                         for rawVersion, updates := range expectedPlugin.FieldUpdateChecks {
				parsedVersion := versioning.MustNewRange(rawVersion)
				config, err := json.ProtoJSONMarshal(dpPluginsMap[name])
				require.NoError(t, err, fmt.Sprintf("marshal %s plugin config: %v", name, err))
				configStr := string(config)
				if parsedVersion(dataPlaneVersion) {
					for _, update := range updates {
						res := gjson.Get(configStr, update.Field)
						if !res.Exists() || res.Value() != update.Value {
							return fmt.Errorf("expected %s, but got %s", update.Value, res.Value())
						}
					}
				}
			}
```

Reply:
You should be able to use [require.JSONEq](https://pkg.go.dev/github.com/stretchr/testify/require#JSONEq) to compare the JSON component instead of the objects themselves.

Reply:
Thanks for the feedback. Looks like the solution with require.JSONEq will require creating of test plugin config fixtures to compare against. I'm going to stick with the approved solution, and we can revisit this if adding checks on types becomes unwieldy to maintain 😄